gtsfm/averaging/rotation/shonan.py

```diff
@@ -30,20 +30,33 @@ class ShonanRotationAveraging(RotationAveragingBase):
     """Performs Shonan rotation averaging."""
 
     def __init__(self) -> None:
+        """
+        Note: `p_min` and `p_max` describe the minimum and maximum relaxation rank.
+        """
         self._p_min = 5
         self._p_max = 30
 
-    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
-        """Run the rotation averaging.
+    def run_w_consecutive_ordering(
+        self, num_connected_nodes: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
+    ) -> List[Optional[Rot3]]:
+        """Run the rotation averaging on a connected graph w/ N keys ordered consecutively [0,...,N-1].
+
+        Note: the assumption that Shonan keys are ordered as [0,...,N-1] goes deeper than
+        ShonanAveraging::NrUnknowns(), which just checks the assumption. Several other functions that construct
+        matrices related to the convergence certificate also use the fact that the graph is connected and
+        keys are 0..N-1. There are about 10 places where nrUnknowns is used to dimension a sparse or dense matrix.
+        Modifying GTSAM would require a major philosophical overhaul, so we perform the re-ordering
+        here in a sort of "wrapper". See: https://github.com/borglab/gtsam/issues/784
 
         Args:
-            num_images: number of poses.
-            i2Ri1_dict: relative rotations as dictionary (i1, i2): i2Ri1.
+            num_connected_nodes: number of unique connected nodes in the graph (can be far less than the number
+                of images in the dataset)
+            i2Ri1_dict: relative rotations for each image pair-edge as dictionary (i1, i2): i2Ri1.
```

Comment:
Do both i1 and i2 have to be less than num_connected_nodes? Could we document that? Do these edges have to form a connected component? Can there be "extra" edges?

Reply:
@akshay-krishnan  So technically Shonan will still return results if the graph is disconnected. For example, if we give i1Ri0, and i3Ri2, Shonan will give back 4 absolute rotations. But they won't make sense, e.g.
```
[R: [
	-0.00275314, -0.883536, 0.468356;
	-0.605804, -0.371157, -0.703735;
	0.795609, -0.285669, -0.534228
]
, R: [
	-0.00275314, -0.883536, 0.468356;
	-0.605804, -0.371157, -0.703735;
	0.795609, -0.285669, -0.534228
]
, R: [
	-0.112206, -0.929598, 0.351081;
	-0.94733, 0.206725, 0.244601;
	-0.299958, -0.305144, -0.903832
]
, R: [
	-0.112206, -0.929598, 0.351081;
	-0.94733, 0.206725, 0.244601;
	-0.299958, -0.305144, -0.903832
]
]
```

Reply:
```python
        wTi0 = Pose3(Rot3(), np.array([1, 1, 0]))
        wTi1 = Pose3(Rot3(), np.array([3, 1, 0]))
        wTi2 = Pose3(Rot3(), np.array([3, 3, 0]))
        wTi3 = Pose3(Rot3(), np.array([1, 3, 0]))

        # generate i2Ri1 rotations
        # (i1,i2) -> i2Ri1
        i2Ri1_input = {
            (0, 1): wTi1.between(wTi0).rotation(),
            (2, 3): wTi3.between(wTi2).rotation(),
        }
```